gdal/ogr/ogrsf_frmts/carto/ogrcartotablelayer.cpp

```diff
@@ -631,14 +857,199 @@ OGRErr OGRCARTOTableLayer::ICreateFeature( OGRFeature *poFeature )
                 json_object_put(poObj);
         }
     }
+    
+    if (bCopyMode)
+        return ICreateFeatureCopy(poFeature, bHasUserFieldMatchingFID, bHasJustGotNextFID);
+    else
+        return ICreateFeatureInsert(poFeature, bHasUserFieldMatchingFID, bHasJustGotNextFID);
+}
+
+/************************************************************************/
+/*                           ICreateFeatureCopy()                       */
+/************************************************************************/
+
+OGRErr OGRCARTOTableLayer::ICreateFeatureCopy( OGRFeature *poFeature, 
+    bool bHasUserFieldMatchingFID, bool bHasJustGotNextFID )
+{
+    CPLString osCopyFile;
+    GetLayerDefn();
+    
+    /* We are doing a new COPY for each full buffer, so we will */
+    /* construct a new COPY statement here, even though we could */
+    /* reuse the same one over and over if we cached it (hmm) */
+    if( eDeferredInsertState == INSERT_UNINIT )
+    {
+        osCopySQL.clear();
+        osCopySQL.Printf("COPY %s ", OGRCARTOEscapeIdentifier(osName).c_str());
+        bool bMustComma = false;
+        /* Non-spatial column names */
+        for( int i = 0; i < poFeatureDefn->GetFieldCount(); i++ )
+        {
+            if( bMustComma )
+                osCopySQL += ",";
+            else
+            {
+                osCopySQL += "(";
+                bMustComma = true;
+            }
+
+            osCopySQL += OGRCARTOEscapeIdentifier(poFeatureDefn->GetFieldDefn(i)->GetNameRef());
+        }
+        /* Geometry column names */
+        for( int i = 0; i < poFeatureDefn->GetGeomFieldCount(); i++ )
+        {
+            if( bMustComma )
+                osCopySQL += ",";
+            else
+                bMustComma = true;
+
+            osCopySQL += OGRCARTOEscapeIdentifier(poFeatureDefn->GetGeomFieldDefn(i)->GetNameRef());
+        }
+        /* FID column */
+        if( !bHasUserFieldMatchingFID &&
```

Comment:
what about if the first feature copied has a FID already set, and not the next one ? or the reverse

Reply:
Unlike the per-feature insert mode I don't have the freedom to shift the structure if the structure of the input feature stream changes... Similar question could be asked about fields existing or not, I think. I do assume the feature stream is homogeneous (mostly a safe choice). Save the initial state and make noise if the incoming features don't match?

Reply:
I would say that if we detect that the structure changes, we should flush the existing buffer, and start a new copy with the new structure. This is a perfectly valid use case from the point of view of the GDAL API to have different missing fields .
Looking at the PG driver, for regular fields, it uses Copy even in case of unset fields, but only if the field has no default value, otherwise it reverts to Insert ( https://github.com/OSGeo/gdal/blob/master/gdal/ogr/ogrsf_frmts/pg/ogrpgtablelayer.cpp#L1526 )

Reply:
The existing CARTO code doesn't seem to flush on unset fields, once it gets into multi-insert mode, it just puts in a NULL. The first feature in seems to control whether that field is included in the insert or not. That seems like a reasonable(ish) thing to do, as I do expect a certain amount of homogeneity in the stream. If the stream is non-homogeneous, the current insert code will also break (it'll set up an INSERT statement with fewer fields than end up in the VALUES, if the first Feature lacks a field and subsequence features have that field. I'm not sure if my approach (fill in all fields in the COPY setup, and then turn missing fields into NULL) is better or worse than the INSERT behaviour (only fill in fields that are set in the first feature, and then cause an error if you run into subsequence features that do have those fields). The fact that we haven't seen the latter case reported as a bug seems to indicate it's a little rare?

Reply:
ok, so perhaps add some comment in the code to document the current limitations/hypothesis